File: Stack/84. Largest Rectangle in Histogram/Solution.java
import java.util.*;
import java.lang.Math;

class Solution {
    public int largestRectangleArea(int[] heights) {
        // N : len(heights)
        // Time Complexity: O(N^2)
        // Space Complexity: O(1)
        // int n = heights.length;
        // int maxArea = 0;
        // for (int i = 0; i < n; i++) {
        //     int minHeight = heights[i];
        //     for (int j = i; j < n; j++) {
        //         minHeight = Math.min(minHeight, heights[j]);
        //         maxArea = Math.max(maxArea, minHeight * (j - i + 1));
        //     }
        // }
        // return maxArea;


        // N : len(heights)
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        int n = heights.length;
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0;
        for (int i = 0; i < n; i++){
            while (!stack.isEmpty() && heights[stack.peek()] > heights[i]) { 
                int idx = stack.pop();
                int prev = (stack.isEmpty()) ? -1 : stack.peek();
                maxArea = Math.max(maxArea, heights[idx] * (i - prev - 1));
            }
            stack.push(i);
        }

        while (!stack.isEmpty()) {
            int idx = stack.pop();
            int prev = (stack.isEmpty()) ? -1 : stack.peek();
            maxArea = Math.max(maxArea, heights[idx] * (n - prev - 1));
        }
        return maxArea;
    }
}

File: Stack/84. Largest Rectangle in Histogram/Solution.py
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # N : len(heights)
        # Time Complexity: O(N^2)
        # Space Complexity: O(1)
        # n = len(heights)
        # maxArea = 0
        # for i in range(n):
        #     minHeight = heights[i]
        #     for j in range(i, n):
        #         minHeight = min(minHeight, heights[j])
        #         maxArea = max(maxArea, (j - i + 1) * minHeight)
        # return maxArea


        # N : len(heights)
        # Time Complexity: O(N)
        # Space Complexity: O(N)
        n = len(heights)
        stack = []
        maxArea = 0
        for i in range(n):
            while stack and heights[stack[-1]] > heights[i]:
                idx = stack.pop()
                prev = -1 if not stack else stack[-1]
                maxArea = max(maxArea, heights[idx] * (i - prev - 1))
            stack.append(i)
        
        while stack:
            idx = stack.pop()
            prev = -1 if not stack else stack[-1]
            maxArea = max(maxArea, heights[idx] * (n - prev - 1))
        return maxArea